### ISSUE TITLE
Metrics strategy, discussion

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#2465]
I want to start a discussion about metrics. The challenge with metrics is that it is a very open ended topic, and every individual and team seems to have slightly varying monitoring needs and service level objectives. We see a consistent stream of requests for new metrics, some that (for all I know) are only useful to the team requesting them. This concerns me a bit.

I am inclined to suggest that perhaps we need to move metrics outside of Drone to a prometheus exporter so that each team has the ability to fork / customize as needed. Or __alternatively__, perhaps we could agree on a baseline set of metrics included in Drone core (runtime/memory/cpu/gc/http), and then provide a prometheus exporter that enables expanded metrics (repo/build/user/etc) that can be forked and customized by teams. _Note that if we end up taking this approach, I will create and host the exporter as an official drone project_.

I believe that when it comes to metrics, we will never satisfy everyone, so we should instead make it easier to extend and customize Drone. I realize there are some competing interests at play here (features vs code maintainability vs ease of installation etc) so I am hoping we can find some sort of reasonable middle-ground. I would love to get suggestions / ideas from everyone so that we can come up with a strategy and execute for 1.0

**Comment by tboerger** [comment]
Since we got to fight with some extended expvar middleware to get runtime, memory and so on exports via an api it sounds like a useful solution to provide simple basic metrics within drone, and to provide an separate exporter for other metrics. The separate exporter can be forked by anybody and expanded to the needs of any team without polluting the drone core.

**Comment by gtaylor** [comment]
> Or alternatively, perhaps we could agree on a baseline set of metrics included in Drone core (runtime/memory/cpu/gc/http), and then provide a prometheus exporter that enables expanded metrics (repo/build/user/etc) that can be forked and customized by teams.

There is plenty of precedent for this approach, and it sounds like a very reasonable way to go. Kubernetes has done something similar, FWIW (ie: kube-state-metrics).

It's also nice to avoid the extra setup work of standing up an exporter if all you want is the base-level metrics.

**Comment by tboerger** [comment]
Ping @metalmatze since he's pretty good on the prometheus stuff ;)

**Comment by tonglil** [comment]
I wouldn't mind running an exporter for expanded metrics.

One thing to consider is that maintaining the exporter, it might be ideal to allow some "configuration" of the exporter to "turn on" the metrics folks may contribute to it.
Otherwise I can see there being a bunch of forks of the exporter implementing the same/similar metrics but in different combinations, and the work never returns upstream.

**Comment by bradrydzewski** [comment]
@tonglil good point. I will have to research how the kube-state-metrics project handles this. Perhaps we can borrow their approach.

**Comment by tonglil** [comment]
The main benefit of externalizing the "metrics exporter" into a separate component is untying the development velocity of exposing metrics from Drone core.

Querying the DB directly makes it possible without developing a complex API for Drone to transform DB values into API values.

**Comment by tboerger** [comment]
I personally would prefer an api approach instead of direct db access, but maybe this is just me.

**Comment by gtaylor** [comment]
I'd rather maintain a fork with our own metrics in it than expose direct DB access to anything beyond drone-server. This is a fairly sensitive DB. 

I don't think developer velocity is an acceptable reason to take on added surface area to secure and observe. We can get better at managing and evolving the API's development, getting us some of the velocity improvements without creating more risk on the security front.

**Comment by metalmatze** [comment]
I second what @tboerger and @gtaylor said.
Personally I think that Drone itself should be able to provide with all necessary metrics, but I can see that having an exporter might be easier. I only want to point out, that I'd like to have very basic metrics on Drone as well, like goroutines etc. Also, when Drone fails Prometheus can alert me that the target is down.

**Comment by tonglil** [comment]
I posit this then: why build and maintain a metrics API and then build and integrate it w/ a separate API-to-Prometheus translator, instead of just counting the metrics with the Prometheus client library directly and exposing them?

Prometheus text format _is_ a metrics API.

**Comment by gtaylor** [comment]
> why build and maintain a metrics API (that doesn't currently exist) and then build and integrate it w/ a separate API-to-Prometheus translator, instead of just counting the metrics with the Prometheus client library directly and exposing them?

@tonglil Brad had this to say in the issue body above:

> I want to start a discussion about metrics. The challenge with metrics is that it is a very open ended topic, and every individual and team seems to have slightly varying monitoring needs and service level objectives. We see a consistent stream of requests for new metrics, some that (for all I know) are only useful to the team requesting them. This concerns me a bit.

If I am understanding correctly, this sounds like you'd rather continue making the metrics available via Drone's built-in Prometheus exporters. But I did want to take a moment to point out the original prompt.

**Comment by tboerger** [comment]
> I agree with @tonglil. If you have to spend the time creating and maintaining a new API in drone core as well then creating another application to consume the API and repost the data using the prometheus format why not just skip the API and post the data in drone core using the prometheus format.

As already explained above, all the people prefer different metrics or metric formats which seems to be not working for the drone core. So we will build an API that can provide the required information (if this is not already the case) for the different metrics requirements. We provide an exporter that covers the majority which can be easily forked by anybody and extended to the personal needs.

That way everybody can get what he wants or needs without polluting the Drone core with the different requirements.

And if you got super special requirements to the exporter, nobody said you are not allowed to extend the exporter to get direct access to the database.

> While an API could support some sort of query language but these seems overly complex.

Maybe we have to reconsider to provide a GraphQL endpoint like @bradrydzewski already experimented with :)

**Comment by msuterski** [comment]
> As already explained above, all the people prefer different metrics or metric formats which seems to be not working for the drone core. So we will build an API that can provide the required information (if this is not already the case) for the different metrics requirements. We provide an exporter that covers the majority which can be easily forked by anybody and extended to the personal needs.

It be helpful if it was stated here what are the "special" cases that would be outside of what drone will provide by default in the exporter, or what would it be that it will provide.

There isn't that many possible metrics/numbers that drone generates and exposing all of them doesn't seem like a big deal.

I'm also not entirely clear how building in an API is less of "polluting of drone core" than adding metrics to the prometheus metrics exporter that drone already uses. But that's just an implementation detail.

As a customer, in general, I do not like a treatment of `we give you an API, you figure it out`, if I want to have visibility into a metric that is not exposed by a product by default. I do not want to maintain my forked/own exporter just because a metric I'm looking for was not in a default set. Why not expose all of them and not require people to fork anything? We use tools to make our lives easier, not to generate more work for ourselves.

Also, format is for the most part just a representation of the data. It's easy to translate between or even support multiple of them if there is a demand from the customers.

Please, think about your customers and how they will be using your product, not if your sourcecode will get ugly. It doesn't matter at the end. You're building a product, not a sourcecode artifact. 

I want a simple solution that works and doesn't require me to jump over hoops to get something done.

**Comment by tonglil** [comment]
@gtaylor:

> If I am understanding correctly, this sounds like you'd rather continue making the metrics available via Drone's built-in Prometheus exporters.

I'm fine with that.

> Or perhaps you are seeing insufficient value for the external exporter.

I see insufficient value for the external exporter if it needs an additional layer in Drone core to be maintained and made compatible. The **development velocity** of the existing Drone API is already limited and adding more "features" to maintain and upgrade between releases will only extend release cycles.

I actually like the external exporter approach if it was essentially the metrics component _separated_ from the core, which allows a fast development velocity to quickly expose information that can be gathered based on some simple DB queries.

---

@tboerger:

> As already explained above, all the people prefer different metrics or metric formats which seems to be not working for the drone core.

Personally, I'm not seeing what the issue is with exposing all of these metrics with Prometheus. Again, its text format is a metrics API (https://vimeo.com/173704322 @ 9:00) that is widely parsed, and can be easily transformed to almost any output destination with a simple tool like [Telegraf](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/prometheus).

> So we will build an API that can provide the required information (if this is not already the case) for the different metrics requirements. We provide an exporter that covers the majority which can be easily forked by anybody and extended to the personal needs.

By providing an API, it's a layer that no one can consume directly and will have to build transformers/exporters into their own systems; where as outputting Prometheus text format means if you want it in CloudWatch/Datadog/Graphite, just configure Telegraf to push it there..

> That way everybody can get what he wants or needs without polluting the Drone core with the different requirements.

The metrics coming out of Drone shouldn't be "aggregated" or "pre-processed" but rather raw counters and histograms that people can slice-and-dice to their own requirements.

---

In summary, why write an API for `api/metrics/build_count_total` or `api/metrics/build_time&unit=s&buckets=10` when you can just integrate the Prometheus client and let that count and expose them for you?

**Comment by gtaylor** [comment]
It sounds like it may be time to pause and get some examples from @bradrydzewski as far as the contention we're trying to relieve. Overall sentiment above seems to be "is the current method really broken?"

**Comment by tonglil** [comment]
To clarify what I mean by using the Prometheus exposition format as the "API": https://www.cncf.io/blog/2018/08/10/cncf-to-host-openmetrics-in-the-sandbox/

**Comment by msuterski** [comment]
@bradrydzewski is there any update/progress on the metrics subject?

**Comment by bradrydzewski** [comment]
after much thought, we will continue build metrics directly into Drone. We have made changes in 1.0 to make this a bit easier. We will evaluate each request for a new metric independently. There will be some requirement that a metric can only be integrated if there is interest from N members of the community to avoid creating one-off metrics for individual teams. I am not sure what N will be, or how it will be measured yet, but I will publish reasonable guidelines once established and agreed upon.
